packages/dashboard-message-bus-client/lib/client.ts

```diff
@@ -0,0 +1,223 @@
+import { DashboardMessageBusClientOptions } from "./types";
+import { createMessage, Message } from "@truffle/dashboard-message-bus";
+
+import { DashboardMessageBusConnection } from "./connection";
+interface SendArgs<M extends Message> {
+  type: M["type"];
+  payload: M["payload"];
+  bestEffort?: boolean;
+}
+
+export class MessageBusConnectionError extends Error {
+  cause: Error | undefined;
+
+  constructor({ message, cause }: { message: string; cause?: Error }) {
+    super(message);
+    this.cause = cause;
+  }
+  toString(): string {
+    return this.cause
+      ? `${super.toString()}\ncaused by: ${this.cause.toString()}`
+      : super.toString();
+  }
+}
+
+export class DashboardMessageBusClient {
+  private _options: Required<DashboardMessageBusClientOptions>;
+
+  private _publishSocket: DashboardMessageBusConnection;
+  private _subscribeSocket: DashboardMessageBusConnection;
+
+  public get options(): Required<DashboardMessageBusClientOptions> {
+    return { ...this._options };
+  }
+
+  //TODO: public settable options that don't require extra steps to take effect
+  public constructor(options: DashboardMessageBusClientOptions) {
+    this._options = {
+      host: "localhost",
+      port: 24012,
+      connectRetryCount: 4,
+      connectRetryDelayMsec: 1000,
+      ...(options ?? {})
+    };
+
+    this._publishSocket = new DashboardMessageBusConnection({
+      host: this._options.host,
+      port: this._options.port,
+      socketType: "publish"
+    });
+
+    this._subscribeSocket = new DashboardMessageBusConnection({
+      host: this._options.host,
+      port: this._options.port,
+      socketType: "subscribe"
+    });
+  }
+
+  async send<MessageType extends Message, ResponseType extends Message>({
+    type,
+    payload,
+    bestEffort
+  }: SendArgs<MessageType>): Promise<ResponseType> {
+    return _withRetriesAsync(
+      () => {
+        const message = createMessage(type, payload);
+        return this._publishSocket.send(message);
+      },
+      this._options.connectRetryCount,
+      bestEffort
+    );
+  }
+
+  async sendAndWait<MessageType extends Message, ResponseType extends Message>({
+    type,
+    payload,
+    bestEffort
+  }: SendArgs<MessageType>): Promise<ResponseType> {
+    return _withRetriesAsync(
+      () => {
+        const message = createMessage(type, payload);
+        return this._publishSocket.sendAndWait<MessageType, ResponseType>(
+          message as MessageType
+        );
+      },
+      this._options.connectRetryCount,
+      bestEffort
+    );
+  }
+
+  async getMessage<ResponseType extends Message>(
```

Comment:
This method's name wasn't clear to me at first what it does. It's also very polymorphic... can we avoid that by giving it its own Options type?